File: python/137.Single Number II.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length=len(nums)
        result=0
        for i in range(32):
        	count=0
        	mask=1<<i
        	for j in range(length):
        		if (nums[j]&mask):
        			count+=1            #统计在某一位上有多少个1
        	if (count%3==1):
        		result|=mask
       	return self.convert(result)

    def convert(self,x):                #python有点坑....
    	if x>=2**31:
    		x-=2**32
    	return x

print(Solution().singleNumber([-2,-2,1,1,-3,1,-3,-3,-4,-2]))

File: cpp/137.Single Number II.cpp
#include <iostream>
#include <vector>

using namespace std; 

class Solution {
	public:
	    int singleNumber(vector<int>& nums) {
	    	int count[32] = {0};
	    	int result = 0;
	    	int n = nums.size();
	        
	        for (int i=0; i<32; i++){
	        	for (int j=0; j<n; j++){
	        		count[i] += ((nums[j]>>i)&1);
	        		count[i] %= 3;
				}
			} 
			for (int i=0; i<32; i++) 
				result += (count[i]<<i);    //用|=速度要快些，但对我而言不太好理解 
			return result; 
	    }   
};

int main(){
	Solution solution = Solution();
	vector<int>nums = {2, 2, 3, 2};
	cout << solution.singleNumber(nums)<<endl;
	
}
